private/pkg/execext/execext.go

```diff
@@ -0,0 +1,250 @@
+func WithStdin(stdin io.Reader) RunStartOption {
+	return &stdinOption{stdin: stdin}
+}
+
+// WithStdout returns a new option that sets the stdout.
+//
+// The default is a [io.Writer] that ignores all writes..
+func WithStdout(stdout io.Writer) RunStartOption {
+	return &stdoutOption{stdout: stdout}
+}
+
+// WithStderr returns a new option that sets the stderr.
+//
+// The default is a [io.Writer] that ignores all writes..
+func WithStderr(stderr io.Writer) RunStartOption {
+	return &stderrOption{stderr: stderr}
+}
+
+// WithDir returns a new option that sets the working directory.
+//
+// The default is the current working directory.
+func WithDir(dir string) RunStartOption {
+	return &dirOption{dir: dir}
+}
+
+// *** PRIVATE ***
+
+type argsOption struct {
+	args []string
+}
+
+func (a *argsOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.args = a.args
+}
+
+func (a *argsOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.args = a.args
+}
+
+type envOption struct {
+	env []string
+}
+
+func (e *envOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.env = e.env
+}
+
+func (e *envOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.env = e.env
+}
+
+type stdinOption struct {
+	stdin io.Reader
+}
+
+func (i *stdinOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.stdin = i.stdin
+}
+
+func (i *stdinOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.stdin = i.stdin
+}
+
+type stdoutOption struct {
+	stdout io.Writer
+}
+
+func (o *stdoutOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.stdout = o.stdout
+}
+
+func (o *stdoutOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.stdout = o.stdout
+}
+
+type stderrOption struct {
+	stderr io.Writer
+}
+
+func (r *stderrOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.stderr = r.stderr
+}
+
+func (r *stderrOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.stderr = r.stderr
+}
+
+type dirOption struct {
+	dir string
+}
+
+func (d *dirOption) applyRun(runStartOptions *runStartOptions) {
+	runStartOptions.dir = d.dir
+}
+
+func (d *dirOption) applyStart(runStartOptions *runStartOptions) {
+	runStartOptions.dir = d.dir
+}
+
+type runStartOptions struct {
+	args   []string
+	env    []string
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+	dir    string
+}
+
+func newRunStartOptions() *runStartOptions {
+	return &runStartOptions{}
+}
+
+func (rs *runStartOptions) applyCmd(cmd *exec.Cmd) {
+	// If the user did not specify env vars, we want to make sure
+	// the command has access to none, as the default is the current env.
+	if len(rs.env) == 0 {
+		cmd.Env = emptyEnv
+	} else {
+		cmd.Env = rs.env
+	}
+	// If the user did not specify any stdin, we want to make sure
+	// the command has access to none, as the default is the default stdin.
+	if rs.stdin == nil {
+		cmd.Stdin = discardReader{}
```

Comment:
Any reason to prefer this over setting `cmd.Stdin = nil` (which will have it read from os.DevNull)?

Reply:
No there's not - good catch and we could change this. I am virtually certain there _was_ a reason, given what I did with stdout and stderr, perhaps os/exec changed the behavior of the default stdin at some point in the last couple of years? I don't know, but I'll confirm that this is true back to 1.22 and then change.

Reply:
This was the case since at least 2014 it turns out.

I'm actually just going to keep this as-is, mostly because I'm terrified of changing it. I agree that there should be no difference to `cmd.Stdin = nil`, but I am somewhat sure I did this for a reason, I can't remember it, and changing it really worries me.

I'm adding a code comment to this effect so that we don't touch this in the future.

Reply:
Comment added.

Reply:
Comment for posterity:

```
// Note: This *should* be the same as just having cmd.Stdin = nil, given that
// exec.Cmd documents that Stdin has the same behavior as Stdout/Stderr, namely
// that os.DevNull is used. This has been the case for Stdin since at least 2014.
// However, way back when this package was first introduced, we set up the discardReader
// for some reason, and we can't remember why. We're terrified to change it, as there
// *may* have been some reason to do so. os.DevNull is actually just a string such
// as "/dev/null" on Unix systems, so how Golang actually handles this is somewhat
// unknown. Honestly, I might just want to change Stdout/Stderr to use a discardWriter.
```